File: Android/app/src/main/java/com/openmediabridge/lyrics/TranslationService.kt
package com.openmediabridge.lyrics

import com.openmediabridge.model.LyricsLine
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Translates lyric lines via a LibreTranslate-compatible endpoint.
 * Mirrors TranslationService.cs - all lines sent in a single request
 * (joined by \n) to minimise API calls; original timestamps are kept.
 */
object TranslationService {
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    fun translateLyrics(
        lines: List<LyricsLine>,
        targetLang: String,
        endpoint: String,
        apiKey: String = ""
    ): List<LyricsLine> {
        if (lines.isEmpty()) return emptyList()

        val texts = lines.map { it.text.ifBlank { " " } }
        val joined = texts.joinToString("\n")

        val body = JSONObject().apply {
            put("q", joined)
            put("source", "auto")
            put("target", targetLang)
            put("format", "text")
            put("api_key", apiKey)
        }.toString()

        val url = endpoint.trimEnd('/') + "/translate"
        val request = Request.Builder()
            .url(url)
            .post(body.toRequestBody("application/json".toMediaType()))
            .build()

        val responseJson = httpClient.newCall(request).execute().use { resp ->
            if (!resp.isSuccessful) throw java.io.IOException("Translation request failed: ${resp.code}")
            resp.body?.string() ?: ""
        }

        val translatedText = JSONObject(responseJson).optString("translatedText", "")
        val translatedLines = translatedText.split("\n")

        return lines.mapIndexed { i, line ->
            LyricsLine(
                timeMs = line.timeMs,
                text = if (i < translatedLines.size) translatedLines[i].trim() else line.text
            )
        }
    }
}

File: Core/Services/TranslationService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using OpenMediaBridge.Lyrics.Fetchers;

namespace OpenMediaBridge.Services
{
    public static class TranslationService
    {
        private static readonly HttpClient _httpClient = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(30)
        };

        /// <summary>
        /// Translates a list of lyric lines via a LibreTranslate-compatible endpoint.
        /// All lines are sent in a single request (joined by \n) to minimise API calls.
        /// Original timestamps are preserved on the returned lines.
        /// </summary>
        public static async Task<List<LyricsLine>> TranslateLyricsAsync(
            List<LyricsLine> lines,
            string targetLang,
            string endpoint,
            string apiKey = "")
        {
            if (lines == null || lines.Count == 0)
                return new List<LyricsLine>();

            var texts = lines.Select(l => string.IsNullOrWhiteSpace(l.Text) ? " " : l.Text).ToList();
            var joined = string.Join("\n", texts);

            var requestBody = new TranslateRequest
            {
                Q = joined,
                Source = "auto",
                Target = targetLang,
                Format = "text",
                ApiKey = apiKey ?? ""
            };

            var json = JsonSerializer.Serialize(requestBody);
            var content = new StringContent(json, Encoding.UTF8, "application/json");

            var url = endpoint.TrimEnd('/') + "/translate";
            var response = await _httpClient.PostAsync(url, content);
            response.EnsureSuccessStatusCode();

            var responseJson = await response.Content.ReadAsStringAsync();
            var result = JsonSerializer.Deserialize<TranslateResponse>(responseJson);

            var translatedTexts = result?.TranslatedText?.Split('\n') ?? Array.Empty<string>();

            var output = new List<LyricsLine>(lines.Count);
            for (int i = 0; i < lines.Count; i++)
            {
                output.Add(new LyricsLine
                {
                    Time = lines[i].Time,
                    Text = i < translatedTexts.Length ? translatedTexts[i].Trim() : lines[i].Text
                });
            }
            return output;
        }

        private class TranslateRequest
        {
            [JsonPropertyName("q")]
            public string Q { get; set; }
            [JsonPropertyName("source")]
            public string Source { get; set; }
            [JsonPropertyName("target")]
            public string Target { get; set; }
            [JsonPropertyName("format")]
            public string Format { get; set; }
            [JsonPropertyName("api_key")]
            public string ApiKey { get; set; }
        }

        private class TranslateResponse
        {
            [JsonPropertyName("translatedText")]
            public string TranslatedText { get; set; }
        }
    }
}
